draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1439,159 +1538,248 @@ Steps:
-typical and RECOMMENDED choices.
-
-For example, for 128-bit security, b >= 256 bits; in this case, SHA256 would
-be an appropriate choice for H.
-
-Ensuring that the hash\_to\_base output is a uniform random element of F requires
-care, even when H is modeled as a random oracle. For example,
-if H is SHA256 and F is a field of characteristic p = 2^255 - 19, then the
-result of reducing H(msg) (a 256-bit integer) modulo p is slightly more likely
-to be in \[0, 37\] than if the value were selected uniformly at random.
-In this example the bias is negligible, but in general it can be significant.
-
-To control bias, the input msg should be hashed to an integer comprising at
-least ceil(log2(p)) + k bits; reducing this integer modulo p gives bias at
-most 2^-k, which is a safe choice for a cryptosystem with k-bit security.
-To obtain such an integer, HKDF {{!RFC5869}} is used to expand the input
-msg to a L-byte string, where L = ceil((ceil(log2(p)) + k) / 8); this
-string is then interpreted as an integer via OS2IP {{RFC8017}}. For example,
-for p a 255-bit prime and k = 128-bit security, L = ceil((255 + 128) / 8) = 48 bytes.
-
-Finally, hash\_to\_base appends one zero byte to msg in the invocation of HKDF-Extract.
-This ensures that the use of HKDF in hash\_to\_base is indifferentiable
-from a random oracle (see {{LBB19}}, Lemma 8 and {{DRST12}}, Theorems 4.3 and 4.4).
-(In particular, this approach works because it ensures that the final byte of
-each HMAC invocation in HKDF-Extract and HKDF-Expand is distinct.)
+The reason is that rejection sampling procedures are difficult to implement
+in constant time, and later well-meaning "optimizations" may silently render
+an implementation non-constant-time.
+
+## Security considerations {#hashtofield-sec}
+
+The hash\_to\_field function is designed to be indifferentiable from a
+random oracle {{MRH04}} when expand\_message ({{hashtofield-expand}})
+is modeled as a random oracle (see {{security-considerations-hash-to-field}}).
+Ensuring indifferentiability requires care; to see why, consider a prime
+p that is close to 3/4 * 2^256.
+Reducing a random 256-bit integer modulo this p yields a value that is in
+the range \[0, p / 3\] with probability roughly 1/2, meaning that this value
+is statistically far from uniform in \[0, p - 1\].
+
+To control bias, hash\_to\_field instead uses pseudorandom integers whose
+length is at least ceil(log2(p)) + k bits.
+Reducing such integers mod p gives bias at most 2^-k for any p; this bias
+is appropriate for a cryptosystem with k-bit security.
+To obtain such integers, hash\_to\_field uses expand\_message to obtain
+L pseudorandom octets, where L = ceil((ceil(log2(p)) + k) / 8); this
+octet-string is then interpreted as an integer via OS2IP {{RFC8017}}.
+For example, for p a 255-bit prime and k = 128-bit security,
+L = ceil((255 + 128) / 8) = 48 octets.
+
+## hash\_to\_field implementation {#hashtofield-impl}
+
+The following procedure implements hash\_to\_field.
```

Comment:
somebody ask me whether these methods can be used for small characteristic field. Ex. binary or ternary field extensions.?
Obviously we are not covering binary or ternary curves.

Reply:
Hmmm, this is a great question. I guess it probably works, but I'd bet there are more efficient ways...